File: java-lib/src/main/java/com/android/layoutlib/bridge/intensive/util/SessionParamsBuilder.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.layoutlib.bridge.intensive.util;

import com.android.SdkConstants;
import com.android.ide.common.rendering.api.AssetRepository;
import com.android.ide.common.rendering.api.LayoutLog;
import com.android.ide.common.rendering.api.LayoutlibCallback;
import com.android.ide.common.rendering.api.ResourceNamespace;
import com.android.ide.common.rendering.api.ResourceReference;
import com.android.ide.common.rendering.api.SessionParams;
import com.android.ide.common.rendering.api.SessionParams.RenderingMode;
import com.android.ide.common.resources.ResourceResolver;
import com.android.ide.common.resources.configuration.FolderConfiguration;
import com.android.ide.common.resources.deprecated.ResourceRepository;
import com.android.layoutlib.bridge.intensive.setup.ConfigGenerator;
import com.android.layoutlib.bridge.intensive.setup.LayoutPullParser;
import com.android.resources.ResourceType;

import android.annotation.NonNull;

import java.util.HashMap;
import java.util.Map;

import com.google.common.collect.ImmutableMap;

/**
 * Builder to help setting up {@link SessionParams} objects.
 */
public class SessionParamsBuilder {

    private LayoutPullParser mLayoutParser;
    private RenderingMode mRenderingMode = RenderingMode.NORMAL;
    private Object mProjectKey = null;
    private ConfigGenerator mConfigGenerator = ConfigGenerator.NEXUS_5;
    private ResourceRepository mFrameworkResources;
    private ResourceRepository mProjectResources;
    private String mThemeName;
    private boolean isProjectTheme;
    private LayoutlibCallback mLayoutlibCallback;
    private int mTargetSdk;
    private int mMinSdk = 0;
    private LayoutLog mLayoutLog;
    private Map<SessionParams.Key, Object> mFlags = new HashMap<>();
    private AssetRepository mAssetRepository = null;
    private boolean mDecor = true;

    @NonNull
    public SessionParamsBuilder setParser(@NonNull LayoutPullParser layoutParser) {
        mLayoutParser = layoutParser;

        return this;
    }

    @NonNull
    public SessionParamsBuilder setRenderingMode(@NonNull RenderingMode renderingMode) {
        mRenderingMode = renderingMode;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setConfigGenerator(@NonNull ConfigGenerator configGenerator) {
        mConfigGenerator = configGenerator;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setProjectResources(@NonNull ResourceRepository resources) {
        mProjectResources = resources;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setFrameworkResources(@NonNull ResourceRepository resources) {
        mFrameworkResources = resources;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTheme(@NonNull String themeName, boolean isProjectTheme) {
        mThemeName = themeName;
        this.isProjectTheme = isProjectTheme;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTheme(@NonNull String themeName) {
        boolean isProjectTheme;
        if (themeName.startsWith(SdkConstants.PREFIX_ANDROID)) {
            themeName = themeName.substring(SdkConstants.PREFIX_ANDROID.length());
            isProjectTheme = false;
        } else {
            isProjectTheme = true;
        }
        return setTheme(themeName, isProjectTheme);
    }

    @NonNull
    public SessionParamsBuilder setCallback(@NonNull LayoutlibCallback callback) {
        mLayoutlibCallback = callback;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setTargetSdk(int targetSdk) {
        mTargetSdk = targetSdk;
        return this;
    }

    @SuppressWarnings("unused")
    @NonNull
    public SessionParamsBuilder setMinSdk(int minSdk) {
        mMinSdk = minSdk;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setLayoutLog(@NonNull LayoutLog layoutLog) {
        mLayoutLog = layoutLog;
        return this;
    }

    @NonNull
    public SessionParamsBuilder setFlag(@NonNull SessionParams.Key flag, Object value) {
        mFlags.put(flag, value);
        return this;
    }

    @NonNull
    public SessionParamsBuilder setAssetRepository(@NonNull AssetRepository repository) {
        mAssetRepository = repository;
        return this;
    }

    @NonNull
    public SessionParamsBuilder disableDecoration() {
        mDecor = false;
        return this;
    }

    @NonNull
    public SessionParams build() {
        assert mFrameworkResources != null;
        assert mProjectResources != null;
        assert mThemeName != null;
        assert mLayoutLog != null;
        assert mLayoutlibCallback != null;

        FolderConfiguration config = mConfigGenerator.getFolderConfig();
        ResourceResolver resourceResolver = ResourceResolver.create(
                ImmutableMap.of(
                        ResourceNamespace.ANDROID, mFrameworkResources.getConfiguredResources(config),
                        ResourceNamespace.TODO(), mProjectResources.getConfiguredResources(config)),
                new ResourceReference(
                        ResourceNamespace.fromBoolean(!isProjectTheme),
                        ResourceType.STYLE,
                        mThemeName));

        SessionParams params = new SessionParams(mLayoutParser, mRenderingMode, mProjectKey /* for
        caching */, mConfigGenerator.getHardwareConfig(), resourceResolver, mLayoutlibCallback,
                mMinSdk, mTargetSdk, mLayoutLog);

        mFlags.forEach(params::setFlag);
        params.setAssetRepository(mAssetRepository);

        if (!mDecor) {
            params.setForceNoDecor();
        }

        return params;
    }
}

File: paparazzi/src/main/java/app/cash/paparazzi/internal/SessionParamsBuilder.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.paparazzi.internal

import com.android.SdkConstants
import com.android.ide.common.rendering.api.AssetRepository
import com.android.ide.common.rendering.api.ResourceNamespace
import com.android.ide.common.rendering.api.ResourceReference
import com.android.ide.common.rendering.api.SessionParams
import com.android.ide.common.rendering.api.SessionParams.Key
import com.android.ide.common.rendering.api.SessionParams.RenderingMode
import com.android.ide.common.resources.ResourceResolver
import com.android.ide.common.resources.ResourceValueMap
import com.android.ide.common.resources.deprecated.ResourceRepository
import com.android.resources.ResourceType

/** Creates [SessionParams] objects. */
internal data class SessionParamsBuilder(
  private val layoutlibCallback: PaparazziCallback,
  private val logger: PaparazziLogger,
  private val frameworkResources: ResourceRepository,
  private val assetRepository: AssetRepository,
  private val projectResources: ResourceRepository,
  private val deviceConfig: DeviceConfig = DeviceConfig.NEXUS_5,
  private val renderingMode: RenderingMode = RenderingMode.NORMAL,
  private val targetSdk: Int = 22,
  private val flags: Map<Key<*>, Any> = mapOf(),
  private val themeName: String? = null,
  private val isProjectTheme: Boolean = false,
  private val layoutPullParser: LayoutPullParser? = null,
  private val projectKey: Any? = null,
  private val minSdk: Int = 0,
  private val decor: Boolean = true
) {
  fun withTheme(
    themeName: String,
    isProjectTheme: Boolean
  ): SessionParamsBuilder {
    return copy(themeName = themeName, isProjectTheme = isProjectTheme)
  }

  fun withTheme(themeName: String): SessionParamsBuilder {
    return when {
      themeName.startsWith(SdkConstants.PREFIX_ANDROID) -> {
        withTheme(themeName.substring(SdkConstants.PREFIX_ANDROID.length), false)
      }
      else -> withTheme(themeName, true)
    }
  }

  fun plusFlag(
    flag: SessionParams.Key<*>,
    value: Any
  ) = copy(flags = flags + (flag to value))

  fun build(): SessionParams {
    require(themeName != null)

    val folderConfiguration = deviceConfig.folderConfiguration
    val resourceResolver = ResourceResolver.create(
        mapOf<ResourceNamespace, Map<ResourceType, ResourceValueMap>>(
            ResourceNamespace.ANDROID to frameworkResources.getConfiguredResources(
                folderConfiguration
            ),
            ResourceNamespace.TODO() to projectResources.getConfiguredResources(
                folderConfiguration
            )
        ),
        ResourceReference(
            ResourceNamespace.fromBoolean(!isProjectTheme),
            ResourceType.STYLE,
            themeName
        )
    )

    val result = SessionParams(
        layoutPullParser, renderingMode, projectKey /* for caching */,
        deviceConfig.hardwareConfig, resourceResolver, layoutlibCallback, minSdk, targetSdk, logger
    )

    for ((key, value) in flags) {
      result.setFlag(key as Key<Any>, value)
    }
    result.setAssetRepository(assetRepository)

    if (!decor) {
      result.setForceNoDecor()
    }

    return result
  }
}
